src/libexpr/value.cc

```diff
@@ -0,0 +1,54 @@
+#include "value.hh"
+#include "eval.hh"
+
+namespace nix {
+
+const Value * getPrimOp(const Value &v) {
+    const Value * primOp = &v;
+    while (primOp->isPrimOpApp()) {
+        primOp = primOp->primOpApp.left;
+    }
+    assert(primOp->isPrimOp());
+    return primOp;
+}
+
+std::string_view showType(ValueType type, bool withArticle)
+{
+    #define WA(a, w) withArticle ? a " " w : w
+    switch (type) {
+        case nInt: return WA("an", "integer");
+        case nBool: return WA("a", "Boolean");
+        case nString: return WA("a", "string");
+        case nPath: return WA("a", "path");
+        case nNull: return "null";
+        case nAttrs: return WA("a", "set");
+        case nList: return WA("a", "list");
+        case nFunction: return WA("a", "function");
+        case nExternal: return WA("an", "external value");
+        case nFloat: return WA("a", "float");
+        case nThunk: return WA("a", "thunk");
+    }
+    abort();
+}
+
+std::string showType(const Value & v)
+{
+    // Allow selecting a subset of enum values
+    #pragma GCC diagnostic push
+    #pragma GCC diagnostic ignored "-Wswitch-enum"
```

Comment:
Can we just list the ones we want to combine below with follow-through? There is a attribute thing to say "this follow-through is intentional".